api/desecapi/models.py

```diff
@@ -143,35 +143,35 @@ def sync_to_pdns(self):
         Make sure that pdns gets the latest information about this domain/zone.
         Re-Syncing is relatively expensive and should not happen routinely.
         """
-
         # Create zone if it does not exist yet
         try:
             # This will also import any RRsets that may have been created already
             pdns.create_zone(self)
         except pdns.PdnsException as e:
             if (e.status_code == 422 and e.detail.endswith(' already exists')):
-                # Delete all RRsets from pdns
-                pdns.set_rrsets(self, [])
-
-                # Update zone to latest information
+                # Zone exists, purge it by deleting all RRsets
+                pdns.set_rrsets(self, [], notify=False)
+                # Update pdns zone to latest information
                 pdns.set_rrsets(self, self.get_rrsets())
             else:
                 raise e
 
     @transaction.atomic
     def sync_from_pdns(self):
         RRset.objects.filter(domain=self).delete()
-        for rrset in pdns.get_rrsets(self):
-            if rrset['type'] not in RRset.RESTRICTED_TYPES:
-                RRset(**rrset).save(pdns=False)
-        ## TODO is this better?
-        # rrsets = pdns.get_rrsets(self)
-        # rrsets = [rrset for rrset in rrsets if
-        #           rrset.type not in RRset.RESTRICTED_TYPES]
-        # RRset.objects.bulk_create(rrsets)
+        rrset_datas = [rrset_data for rrset_data in pdns.get_rrset_datas(self)
+                       if rrset_data['type'] not in RRset.RESTRICTED_TYPES]
+        # Can't do bulk create because we need records creation in RRset.save()
+        for rrset_data in rrset_datas:
+            RRset(**rrset_data).save(sync=False)
```

Comment:
potential run time optimization:

- first create all `RRset` objects
- bulk insert all RRset objects
- collect all dependent `RR` objects
- bulk insert all `RR` objects

that should reduce overhead to two DB transactions